File: Softeer/cpp/6294_성적_평균.cpp
#include <iostream>
#include <vector>

using namespace std;

int n, k; //학생 수, 구간 수

double getAvg(int sum, int num) {

	return (double)sum / num;
}

void settingCumulativeSum(vector<int>& score, vector<int>& cumul) {
	cumul[1] = score[1];

	for (int i = 2; i < score.size(); i++) {
		cumul[i] = cumul[i - 1] + score[i];
	}
}

int getSum(int start, int end, vector<int>& cumul) {
	if (start == 1) {
		return cumul[end];
	}

	return cumul[end] - cumul[start - 1];
}

void solution() {
	cin >> n >> k;

	vector<int> score(n + 1);

	for (int i = 1; i <= n; i++) {
		int a;
		cin >> a;
		score[i] = a;
	}

	vector<int> cumul(score.size());

	settingCumulativeSum(score, cumul);

	for (int i = 0; i < k; i++) {
		int start, end;
		cin >> start >> end;

		int sum = getSum(start, end, cumul);

		double answer = getAvg(sum, end - start + 1);

		printf("%.2f\n", answer);
	}
}

int main(int argc, char** argv)
{
	ios::sync_with_stdio(false);

	solution();

	return 0;
}

File: Softeer/java/6294_성적_평균.java
import java.io.*;
import java.util.*;

public class Main {
    
    public static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    public static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

    public static int n, k; //학생 수, 구간 수
    
    
    public static void main(String[] args) throws IOException {
        solution();

        bw.flush();
        bw.close();
    }

    private static void solution() throws IOException {
        StringTokenizer st = new StringTokenizer(br.readLine());

        n = Integer.parseInt(st.nextToken());
        k = Integer.parseInt(st.nextToken());

        int[] score = new int[n + 1];
        st = new StringTokenizer(br.readLine());
        
        int idx = 1;
        for (int i = 0; i < n; i++) {
            score[idx++] = Integer.parseInt(st.nextToken());
        }

        int[] cumul = new int[score.length];

        settingCumulativeSum(score, cumul);
        
        for (int i = 0; i < k; i++) {
            st = new StringTokenizer(br.readLine());

            int start = Integer.parseInt(st.nextToken());
            int end = Integer.parseInt(st.nextToken());

            int sum = getSum(start, end, cumul);

            double answer = getAvg(sum, end - start + 1);

            bw.write(String.format("%.2f", answer) + "\n");
        }
    }

    private static int getSum(int start, int end, int[] cumul) {
        if (start == 1) {
                return cumul[end];
            }
        
        return cumul[end] - cumul[start - 1];
    }
    
    private static void settingCumulativeSum(int[] score, int[] cumul) {
        cumul[1] = score[1];
        
        for (int i = 2; i < score.length; i++) {
            cumul[i] = cumul[i - 1] + score[i];
        }
    }

    private static double getAvg(int sum, int num) {

        return (double)sum / num;
    }
}
